lib/preload.js

```diff
@@ -361,32 +360,67 @@ class Preloader extends EventEmitter {
 
 	_prepareErrorHandler() {
 		// Emit an error event if the python script exits with an error
-		this.container.wait().then((data) => {
-			if (data.StatusCode !== 0 && data.StatusCode !== EXIT_CAUSED_BY_SIGINT) {
-				const output = this.bufferedStderr.getData().toString('utf8').trim();
-				let error;
-				if (
-					output.indexOf(GRAPHDRIVER_ERROR) !== -1 &&
-					output.indexOf(DOCKERD_USES_OVERLAY) !== -1
-				) {
-					error = new this.balena.errors.BalenaError(OVERLAY_MODULE_MESSAGE);
-				} else {
-					error = new Error(output);
-					// @ts-ignore
-					error.code = data.StatusCode;
+		this.container
+			.wait()
+			.then((data) => {
+				if (data.StatusCode !== 0) {
+					const output = this.bufferedStderr.getData().toString('utf8').trim();
+					let error;
+					if (
+						output.indexOf(GRAPHDRIVER_ERROR) !== -1 &&
+						output.indexOf(DOCKERD_USES_OVERLAY) !== -1
+					) {
+						error = new this.balena.errors.BalenaError(OVERLAY_MODULE_MESSAGE);
+					} else {
+						error = new Error(output);
+						// @ts-ignore
+						error.code = data.StatusCode;
+					}
+					this.emit('error', error);
 				}
-				this.emit('error', error);
-			}
-		});
+			})
+			.catch((error) => this.emit('error', error));
 	}
 
+	/**
+	 * Send a command to `preload.py` (running in a Docker container) through
+	 * its stdin, and read a response from its stdout. The stdout response
+	 * must be a JSON object containing a `StatusCode` field (number / integer)
+	 * and an `Error` field (string). If `StatusCode` is the number zero, the
+	 * command execution is considered successful, otherwise it is assumed to
+	 * have failed and the returned promise is rejected with an error message
+	 * including the message provided in the `Error` field (if any).
+	 *
+	 * @param {string} command
+	 * @param {object} parameters
+	 */
 	_runCommand(command, parameters) {
 		return new Bluebird((resolve, reject) => {
+			const cmd = JSON.stringify({ command, parameters }) + '\n';
 			this.stdout.once('error', reject);
 			this.stdout.once('data', (data) => {
-				resolve(JSON.parse(data).result);
+				let result = {};
+				try {
+					result = JSON.parse(data).result;
+				} catch (error) {
+					result.error = error;
+				}
+				if (result.error) {
+					const msg = [
+						`An error has occurred executing internal preload command '${command}':`,
+						cmd,
+					];
+					msg.push(
+						`StatusCode: 1`,
+						`Error: ${result.error}`,
```

Comment:
Here we still want to return `StatusCode` and `Error` for any calling functions that are expecting for dockerode errors in that format.

Reply:
Thanks your comments @klutchell 👍 
I found a few other issues while testing the code -- it was properly broken, not sure how I missed all of it before creating this PR last year. 😬
* I have moved some more code out of this PR to further focus the review into the "interprocess communication" aspect. Besides PR #239, some code was moved to PR #241, also self certified and merged. Feel free to review it post mortem if you have any concerns!
* For the remaining code, I didn't want to overwrite your fixup commit here so I created PR #240, which is looking more promising in my testing. Could you have a look at that PR, and move any comments there if you think they still apply? 🙏